persistence-api/src/main/java/io/stargate/db/cdc/DefaultCDCHealthChecker.java

```diff
@@ -0,0 +1,136 @@
+package io.stargate.db.cdc;
+
+import com.codahale.metrics.Clock;
+import com.codahale.metrics.EWMA;
+import com.datastax.oss.driver.shaded.guava.common.annotations.VisibleForTesting;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+class DefaultCDCHealthChecker implements CDCHealthChecker {
+  private double errorRateThreshold;
+  private int minErrorsPerSecond;
+  static final int INTERVAL = 5;
+  static final long TICK_INTERVAL = TimeUnit.SECONDS.toNanos(INTERVAL);
+
+  private AutoTickEWMA successes;
+  private AutoTickEWMA errors;
+
+  /**
+   * Creates a new instance of {@link DefaultCDCHealthChecker}.
+   *
+   * @param errorRateThreshold The percentage of error requests compared to total, expressed in as a
+   *     double from 0 to 1.
+   * @param minErrorsPerSecond The minimum amount of error occurrences per second for the health
+   *     checker to consider the error ratio. This setting is created to prevent detecting the CDC
+   *     producer as unhealthy when there's low traffic and few errors.
+   * @param ewmaIntervalMinutes The interval to determine the coefficient for the degree of
+   *     weighting decrease in the exponentially weighted moving average (EWMA). The health checker
+   *     will use this value to set a soothing factor equivalent to UNIX load average.
+   */
+  DefaultCDCHealthChecker(
+      double errorRateThreshold, int minErrorsPerSecond, int ewmaIntervalMinutes) {
+    this(errorRateThreshold, minErrorsPerSecond, ewmaIntervalMinutes, TickClock.defaultClock);
+  }
+
+  @VisibleForTesting
+  DefaultCDCHealthChecker(
+      double errorRateThreshold, int minErrorsPerSecond, int ewmaIntervalMinutes, TickClock clock) {
+
+    if (errorRateThreshold <= 0 || errorRateThreshold > 1) {
+      throw new IllegalArgumentException(
+          "Error rate threshold should be greater than 0 and lower than 1");
+    }
+
+    if (ewmaIntervalMinutes <= 0 || ewmaIntervalMinutes > 15) {
+      throw new IllegalArgumentException(
+          "The interval used to determine the smoothing factor for the exponentially "
+              + "weighted moving average must be higher 0 and lower than 15 minutes");
+    }
+
+    this.errorRateThreshold = errorRateThreshold;
+    this.minErrorsPerSecond = minErrorsPerSecond;
+    successes = new AutoTickEWMA(ewmaIntervalMinutes, clock);
+    errors = new AutoTickEWMA(ewmaIntervalMinutes, clock);
+  }
+
+  @Override
+  public boolean isHealthy() {
+    double errorRate = errors.getRate();
+    if (errorRate < minErrorsPerSecond) {
+      return true;
+    }
+
+    double successRate = successes.getRate();
+
+    double percentage = errorRate / (successRate + errorRate);
+    return percentage < errorRateThreshold;
+  }
```

Comment:
Couldn't we leave this to be calculated externally? We just return all the raw data and leave it up to whatever aggregation tool (e.g. grafana) to calculate healthy? This also frees up the user from needing to redeploy to update thresholds

Reply:
We already expose metrics on the CDC service.
The goal of this task is to provide an automatic circuit breaker as [proposed in the doc](https://docs.google.com/document/d/1iU5nTtizmOxBiBj9JChr9IYmmg0_sZUd7tC8qsFiYdc/edit?disco=AAAAG-JbLos).

It's implemented in a way that, after it's marked as unhealthy, it will automatically probe the underlying producer as the time passes (see `shouldConsiderBeBackToHealthyOnceTheTimePasses()` test).

Reply:
Oooohh, got it.